crates/bevy_pbr/src/atmosphere/functions.wgsl

```diff
@@ -0,0 +1,208 @@
+
+    let mie_scattering = mie_density * atmosphere.mie_scattering;
+    let mie_absorption = mie_density * atmosphere.mie_absorption;
+    let mie_extinction = mie_scattering + mie_absorption;
+
+    let rayleigh_scattering = rayleigh_density * atmosphere.rayleigh_scattering;
+    // no rayleigh absorption
+    // rayleigh extinction is the sum of scattering and absorption
+
+    // ozone doesn't contribute to scattering
+    let ozone_absorption = ozone_density * atmosphere.ozone_absorption;
+
+    var sample: AtmosphereSample;
+    sample.rayleigh_scattering = rayleigh_scattering;
+    sample.mie_scattering = mie_scattering;
+    sample.extinction = rayleigh_scattering + mie_extinction + ozone_absorption;
+
+    return sample;
+}
+
+fn sample_local_inscattering(local_atmosphere: AtmosphereSample, transmittance_to_sample: vec3<f32>, view_dir: vec3<f32>, local_r: f32, local_up: vec3<f32>) -> vec3<f32> {
+    //TODO: storing these outside the loop saves several multiplications, but at the cost of an extra vector register
+    var rayleigh_scattering = vec3(0.0);
+    var mie_scattering = vec3(0.0);
+    for (var light_i: u32 = 0u; light_i < lights.n_directional_lights; light_i++) {
+        let light = &lights.directional_lights[light_i];
+        let mu_light = dot((*light).direction_to_light, local_up);
+        let neg_LdotV = dot((*light).direction_to_light, view_dir);
+        let rayleigh_phase = rayleigh(neg_LdotV);
+        let mie_phase = henyey_greenstein(neg_LdotV);
+
+        let transmittance_to_light = sample_transmittance_lut(local_r, mu_light);
+        let shadow_factor = transmittance_to_light * f32(!ray_intersects_ground(local_r, mu_light));
+
+        let psi_ms = sample_multiscattering_lut(local_r, mu_light);
+
+        rayleigh_scattering += (transmittance_to_sample * shadow_factor * rayleigh_phase + psi_ms) * (*light).color.rgb;
+        mie_scattering += (transmittance_to_sample * shadow_factor * mie_phase + psi_ms) * (*light).color.rgb;
+    }
+    return local_atmosphere.rayleigh_scattering * rayleigh_scattering + local_atmosphere.mie_scattering * mie_scattering;
+}
+
+// TRANSFORM UTILITIES
+
+//We assume the `up` vector at the view position is the y axis, since the world is locally flat/level.
+//NOTE: this means that if your bevy world is actually placed on a sphere, this will be wrong.
+fn get_local_up(view_r: f32, displacement: vec3<f32>) -> vec3<f32> {
+    return normalize(vec3(0.0, view_r, 0.0) + displacement);
+}
+
+fn get_local_r(view_r: f32, view_mu: f32, dist: f32) -> f32 {
+    return sqrt(dist * dist + 2.0 * view_r * view_mu * dist + view_r * view_r);
+}
+
+// Convert uv [0.0 .. 1.0] coordinate to ndc space xy [-1.0 .. 1.0]
+fn uv_to_ndc(uv: vec2<f32>) -> vec2<f32> {
+    return uv * vec2(2.0, -2.0) + vec2(-1.0, 1.0);
+}
+
+/// Convert a ndc space position to world space
+fn position_ndc_to_world(ndc_pos: vec3<f32>) -> vec3<f32> {
```

Comment:
This function already exists under that name in `view_transformations.wgsl`

Reply:
I copied it so that the import didn't bring in all the mesh view bindings. Or would that not happen?